__tests__/core/type-system.test.ts

```diff
@@ -33,14 +33,21 @@ type DifferingKeys<ActualT, ExpectedT> = {
 }[keyof ActualT | keyof ExpectedT] &
   string
 
-type InexactErrorMessage<ActualT, ExpectedT> = `Mismatched property: ${DifferingKeys<
-  ActualT,
-  ExpectedT
->}`
+type NotExactErrorMessage<ActualT, ExpectedT> = ActualT extends Record<string, unknown>
+  ? ExpectedT extends Record<string, unknown>
+    ? `Mismatched property: ${DifferingKeys<ActualT, ExpectedT>}`
+    : "Expected a non-object type, but received an object"
+  : ExpectedT extends Record<string, unknown>
+  ? "Expected an object type, but received a non-object type"
+  : "Types are not exactly equal"
+
+type IsExact<T1, T2> = [T1] extends [T2] ? ([T2] extends [T1] ? Exact<T1, T2> : never) : never
```

Comment:
Realized this wasn't working super great for things that weren't object-y, like a string union (e.g., `"a" | "b"`)

Reply:
Ah yes, I see that. Nice, I like it!

This still doesn't give us the exact mismatch on string unions. I'm seeing:

```
Argument of type '"1" | "2" | "7" | "3" | "0" | "5" | "4" | "6" | "8" | "9"' is not assignable to parameter of type '"Types are not exactly equal"'.
  Type '"1"' is not assignable to type '"Types are not exactly equal"'
```

But I'm pretty happy with it as-is and we can iterate on it later.

Reply:
Yeah, it's unfortunate. Unions are tough to handle here. I was hoping to maybe fall back to TS telling us what's missing by having it return `T1` for the argument in position `T2`, but this breaks things when `T2 extends T1`.

I'll keep iterating though! I'm hoping with some TLC we can make this really great, or maybe we'll find a library that does all of this for us.